### PR TITLE
#164691215 Validate User Signup

### PR DESCRIPTION
**Pull request by NicholusMuwonge**
#### What does this PR do?

Add relevant responses to signup fields during signup process

#### Description of Task to be completed?
- on signup, all the valid fields should be filled out otherwise user will be prompted to fill them out
- user is prompted to use the proper email format.
- user is asked to make password 8 characters long and also alphanumeric
- user cant signup with the same username or email twice
- on success a dictionary object of their username and email is returned.

#### How should this be manually tested?
- one is needed to clone this repository and pull this branch
- You can test for browser functionality using postman by running your server and testing out this route. http://127.0.0.1:8000/api/users/

#### valid input may look like:
``` {
  {
  "user":{
    "username": "jeckyll",
    "email": "hyde@dark.com",
    "password": "jakejake7"
  }
} 
````
#### response should there after look like:
``` 
{
    "user": {
        "email":  "hyde@dark.com",
        "username": "jeckyll"
    }
}
```

#### example, if you ran this
``` {
  "user":{
    "username": "jeckyll",
    "email": "hyde@dark.com",
    "password": "jakejake"
  }
} 
````
response should be :  
``` {
    "errors": {
        "password": [
            "password must contain a numeric character"
        ]
    }
}
```
#### What are the relevant pivotal tracker stories?
[https://www.pivotaltracker.com/story/show/164691215](url)

### CODE REVIEW

File: authors/settings.py
Comment (reifred) [inline authors/settings.py]: @NicholusMuwonge. You shouldn't be exposing sensitive data. Consider using environment variables.

File: newenv/Include/Python.h
Comment (reifred) [inline newenv/Include/Python.h]: Please consider adding your virtual environment in the .gitignore file to avoid sending it with your codebase.

**Review by reifred — CHANGES_REQUESTED** [review]

File: authors/settings.py
Comment (NicholusMuwonge) [reply authors/settings.py]: This I am Trying to implement but still not functioning. Let me fix it

File: newenv/Include/Python.h
Comment (NicholusMuwonge) [reply newenv/Include/Python.h]: The env file Ive added to the gitignore if you check but seems not to be ignored.

**Review by richien — CHANGES_REQUESTED** [review]
Hi Nicholus, first of all, this is good progress. I have only noticed a few potential changes besides the changes already requested by Fred, I noticed an incomplete sentence in the first bullet point of your description of the tasks completed.
 Also, in regards to how this should be tested manually, I think that we should deal with how to test this feature in the present and not in the future after merging is done. Therefore I find the references to the staging and production environments on Heroku un-useful for testing this feature. Also, I don't see how I can use unittest to test this feature, perharps you could elaborate on what you mean. 
Lastly, I also noticed that you have 5 commits for this change, the coding convention is that we should have 1 commit per feature; perhaps you could squash those commits into one. Other than that, I have tested the feature locally and it works as you have described, well done on that.

**Review by richien — CHANGES_REQUESTED** [review]
Thank you Nicholus for implementing the feedback. I still do not see how running unit tests can help test this implementation. Perharps you should restrict the tests to using postman as that makes more sense in regards to viewing the responses returned by certain requests.

**Review by richien — APPROVED** [review]

**Review by reifred — APPROVED** [review]

File: authors/apps/authentication/validation.py
Comment (RonKbS) [inline authors/apps/authentication/validation.py]: Nice work @NicholusMuwonge.

Please consider a different approach rather than having a class with one static method. Kindly look at serializers.py for a start.

**Review by RonKbS — CHANGES_REQUESTED** [review]
Kindly include a check for whether an email exists.

File: authors/apps/authentication/validation.py
Comment (NicholusMuwonge) [reply authors/apps/authentication/validation.py]: > Kindly include a check for whether an email exists.

hey @RonKbS this is already handled in the exceptions since the email is a unique field, so when you try to signup twice, you'll get an appropriate response, hence there's a test for it too.


output should look like:
``` 
{
    "errors": {
        "email": [
            "user with this email already exists."
        ],
        "username": [
            "user with this username already exists."
        ]
    }
}
```

Comment (NicholusMuwonge) [reply authors/apps/authentication/validation.py]: Let me also change the validation to the serializers.py as recommended

Comment (RonKbS) [reply authors/apps/authentication/validation.py]: I mean exists generally and not whether it has already been used to register with our application.

File: authors/apps/authentication/serializers.py
Comment (reifred) [inline authors/apps/authentication/serializers.py:59]: Thanks, @NicholusMuwonge, for the work you could as well add an extra validator to check for a username.

**Review by reifred — CHANGES_REQUESTED** [review]

File: tests/test_authentication.py
Comment (RonKbS) [inline tests/test_authentication.py]: Is there a reason for moving this method into the tests?

Comment (NicholusMuwonge) [reply tests/test_authentication.py]: I saw the test fail and I know that the email isnt being verified so I tried verifying it before it was tested . It didnt work so I was to remove it from tests once I found solution on like a final commit

Comment (NicholusMuwonge) [reply tests/test_authentication.py]: let me take it out now

File: authors/apps/authentication/serializers.py
Comment (RonKbS) [reply authors/apps/authentication/serializers.py:59]: @NicholusMuwonge, @reifred's concern was not adressed.
Please do so.

**Review by reifred — APPROVED** [review]
Thanks for working on the changes I requested. Though work on the conflicting files for now.

File: authors/apps/authentication/serializers.py
Comment (NicholusMuwonge) [reply authors/apps/authentication/serializers.py:59]: I have added the username validator @reifred

**Comment by NicholusMuwonge** [comment]
> Thanks for working on the changes I requested. Though work on the conflicting files for now.

I have fixed the conflicts

**Comment by RonKbS** [comment]
@NicholusMuwonge, please follow this format when writing commits 👇 
Screenshot 2019-04-08 at 08 01 44


Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.

**Comment by NicholusMuwonge** [comment]
> @NicholusMuwonge, please follow this format when writing commits 
> Screenshot 2019-04-08 at 08 01 44
> 
> Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.

Sorry about closing the PR, on renaming the branch it automatically closed.
About the commits, I write them following the convention shared it just appears different remotely

**Comment by NicholusMuwonge** [comment]
> > @NicholusMuwonge, please follow this format when writing commits
> > Screenshot 2019-04-08 at 08 01 44
> > Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.
> 
> Sorry about closing the PR, on renaming the branch it automatically closed.
> About the commits, I write them following the convention shared it just appears different remotely

let me fix them now

**Comment by NicholusMuwonge** [comment]
> > @NicholusMuwonge, please follow this format when writing commits
> > Screenshot 2019-04-08 at 08 01 44
> > Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.
> 
> Sorry about closing the PR, on renaming the branch it automatically closed.
> About the commits, I write them following the convention shared it just appears different remotely

I have fixed it @RonKbS

File: tests/test_authentication.py
Comment (RonKbS) [inline tests/test_authentication.py]: Kindly seek @richien's help on having the different test cases in their own files.
In this case they can go into files that already exist (tests for serializers can be in the same files as views if they are both concerning authentication or something similar).

Comment (NicholusMuwonge) [reply tests/test_authentication.py]: @RonKbS  I have redistributed to files that test different features kindly please take a look

File: authors/apps/authentication/serializers.py
Comment (RonKbS) [inline authors/apps/authentication/serializers.py:51]: Please slack Trust regarding whether users can be allowed to add a character such as an underscore or a dash in their username. This .isalnum() method restricts their use but I believe they should be used. The PO is to have a final say regarding this.

File: authors/apps/authentication/tests/test_register.py
Comment (RonKbS) [inline authors/apps/authentication/tests/test_register.py]: Please make sure that lines do not exceed 79 characters. Same for lines 71, 80 and 89.

Comment (NicholusMuwonge) [reply authors/apps/authentication/tests/test_register.py]: I have fixed this

File: authors/apps/authentication/serializers.py
Comment (NicholusMuwonge) [reply authors/apps/authentication/serializers.py:51]: The PO says we should allow only numbers and letters

File: authors/apps/authentication/tests/test_register.py
Comment (RonKbS) [reply authors/apps/authentication/tests/test_register.py]: May I know how you've fixed it because, as seen in the picture, the line is still more than 79 characters long.
Screenshot 2019-04-10 at 14 33 32

Comment (NicholusMuwonge) [reply authors/apps/authentication/tests/test_register.py]: sorry, I hadnt then but I have now

**Review by RonKbS — APPROVED** [review]

**Pull request merged**
